mlir-sycl/lib/Transforms/Inliner.cpp

```diff
@@ -0,0 +1,387 @@
+    assert(SrcNode && "Expecting valid source node");
+    assert(TgtNode && "Expecting valid target node");
+  }
+
+  /// Return the func::FuncOp called by `callOp`.
+  static FunctionOpInterface getCalledFunction(const CallOpInterface &Call) {
+    if (SymbolRefAttr SymAttr = const_cast<CallOpInterface &>(Call)
+                                    .getCallableForCallee()
+                                    .dyn_cast<SymbolRefAttr>())
+      return SymbolTable::lookupNearestSymbolFrom(Call, SymAttr);
+
+    return nullptr;
+  }
+
+  CallOpInterface Call;
+  const CallGraphNode *SrcNode, *TgtNode;
+};
+
+inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS,
+                                     const ResolvedCall &RC) {
+  auto PrintCallable = [&](Operation *Op) {
+    llvm::TypeSwitch<Operation *>(Op)
+        .Case<func::FuncOp>([&](func::FuncOp Op) { OS << Op.getSymName(); })
+        .Case<gpu::GPUFuncOp>([&](gpu::GPUFuncOp Op) { Op.print(OS); })
+        .Default(
+            [&](Operation *) { assert(false && "Unhandled operation kind"); });
+  };
+
+  if (!RC.SrcNode->isExternal()) {
+    OS << "SrcNode: ";
+    PrintCallable(RC.SrcNode->getCallableRegion()->getParentOp());
+    OS << "\n";
+  }
+
+  if (!RC.TgtNode->isExternal()) {
+    OS << "TgtNode: ";
+    PrintCallable(RC.TgtNode->getCallableRegion()->getParentOp());
+    OS << "\n";
+  }
+
+  OS << "Call: ";
+  PrintCallable(ResolvedCall::getCalledFunction(RC.Call));
+  OS << "\n";
+
+  return OS;
+}
+
+/// This class represents a specific call graph SCC.
+class CallGraphSCC {
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &,
+                                       const CallGraphSCC &);
+
+public:
+  CallGraphSCC(llvm::scc_iterator<const CallGraph *> &ParentIterator)
+      : ParentIterator(ParentIterator) {}
+
+  /// Return a range over the nodes within this SCC.
+  std::vector<CallGraphNode *>::iterator begin() { return Nodes.begin(); }
+  std::vector<CallGraphNode *>::iterator end() { return Nodes.end(); }
+
+  /// Reset the nodes of this SCC with those provided.
+  void reset(const std::vector<CallGraphNode *> &NewNodes) { Nodes = NewNodes; }
+
+  /// Remove the given node from this SCC.
+  void remove(CallGraphNode *Node) {
+    auto It = llvm::find(Nodes, Node);
+    if (It != Nodes.end()) {
+      Nodes.erase(It);
+      ParentIterator.ReplaceNode(Node, nullptr);
+    }
+  }
+
+private:
+  std::vector<CallGraphNode *> Nodes;
+  llvm::scc_iterator<const CallGraph *> &ParentIterator;
+};
+
+inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS,
+                                     const CallGraphSCC &SCC) {
+  auto PrintCallable = [&](Operation *Op) {
+    llvm::TypeSwitch<Operation *>(Op)
+        .Case<func::FuncOp>([&](func::FuncOp Op) { OS << Op.getSymName(); })
+        .Case<gpu::GPUFuncOp>([&](gpu::GPUFuncOp Op) { Op.print(OS); })
+        .Default(
+            [&](Operation *) { assert(false && "Unhandled operation kind"); });
+  };
```

Comment:
Can we use `FunctionOpInterface::getName()` here and above?

Reply:
> I'm thinking it might make sense to make this pass part of polygeist instead of the SYCL dialect, as we will be dealing with functions of different kinds.

I gave that a though as well and decided to add it in the SYCL dialect for now given it deals with SYCL calls. 
In principle I agree that an inliner pass doesn't really belong to any dialect in particular and is more a 'tool' pass so, in our case, a cgeist pass. 

I'd like to land this in the SYCL dialect for now. But likely will move it in a separate PR if that is OK with you (and we can discuss this offline in more detail).

Reply:
I have used `FunctionOpInterface::getName()` and remove the need for the `TypeSwitch`. Thanks for the suggestion Vic.!

Reply:
I'm OK with adding this pass here. I don't have a strong preference, although I feel a bit more inclined towards the other option. I'm fine with this.